plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/mutate/MutateProcessor.java

```diff
@@ -0,0 +1,312 @@
+    private final List<String> lowercase;
+
+    public MutateProcessor(Map<String, Object> update,
+                           Map<String, String> rename,
+                           Map<String, String> convert,
+                           Map<String, String> split,
+                           Map<String, List<String>> gsub,
+                           Map<String, String> join,
+                           List<String> remove,
+                           List<String> strip,
+                           List<String> uppercase,
+                           List<String> lowercase) {
+        this.update = update;
+        this.rename = rename;
+        this.convert = convert;
+        this.split = split;
+        this.gsub = gsub;
+        this.join = join;
+        this.remove = remove;
+        this.strip = strip;
+        this.uppercase = uppercase;
+        this.lowercase = lowercase;
+    }
+
+    @Override
+    public void execute(Data data) {
+        doUpdate(data);
+        doRename(data);
+        doConvert(data);
+        doSplit(data);
+        doGsub(data);
+        doJoin(data);
+        doRemove(data);
+        doStrip(data);
+        doUppercase(data);
+        doLowercase(data);
+    }
+
+    private void doUpdate(Data data) {
+        if (update == null) { return; }
+
+        for(Map.Entry<String, Object> entry : update.entrySet()) {
+            data.addField(entry.getKey(), entry.getValue());
+        }
+    }
+
+    private void doRename(Data data) {
+        if (rename == null) { return; }
+
+        for(Map.Entry<String, String> entry : rename.entrySet()) {
+            Object oldVal = data.getProperty(entry.getKey());
+            data.addField(entry.getValue(), oldVal);
+        }
+    }
+
+    private Object parseValueAsType(Object oldVal, String toType) {
+        if("integer".equals(toType)) {
+            oldVal = Integer.parseInt(oldVal.toString());
+        } else if("float".equals(toType)) {
+            oldVal = Float.parseFloat(oldVal.toString());
+        } else if("string".equals(toType)) {
+            oldVal = oldVal.toString();
+        } else if("boolean".equals(toType)) {
+            String val = oldVal.toString();
+            if (val.equals("true") || val.equals("t") || val.equals("yes") || val.equals("y") || val.equals("1")) {
+                oldVal = true;
+            } else if (val.equals("false") || val.equals("f") || val.equals("no") || val.equals("n") || val.equals("0")) {
+                oldVal = false;
+            } else {
+                throw new IllegalArgumentException("Cannot convert value: " + oldVal.toString() + "into a boolean");
+            }
+        }
+        return oldVal;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void doConvert(Data data) {
+        if (convert == null) { return; }
+
+        for(Map.Entry<String, String> entry : convert.entrySet()) {
+            String toType = entry.getValue();
+
+            Object oldVal = data.getProperty(entry.getKey());
```

Comment:
this is a good point.

If a "string" value is null, and convert wishes to convert it to an "integer", will it matter for when this object is converted to a document for indexing? like, will it just be equivalent to a `null` in json, making its type arbitrary? Even though it is null, it may have a type associated with it.